cmd/snap-update-ns/change.go

```diff
@@ -59,26 +60,143 @@ var changePerform func(*Change) ([]*Change, error)
+		type tryCreateFlag int
+		const createSymlinks tryCreateFlag = 1
+
+		tryCreate := func(flags tryCreateFlag) error {
+			var err error
+			switch kind {
+			case "":
+				err = secureMkdirAll(path, mode, uid, gid)
+			case "file":
+				err = secureMkfileAll(path, mode, uid, gid)
+			case "symlink":
+				if flags&createSymlinks == createSymlinks {
+					err = secureMkdirAll(path, mode, uid, gid)
+					if err == nil {
+						// Normally symlinks would be created
+						// later but calling c.lowLevelPerform
+						// now lets us check if the medium is
+						// writable and take advantage of the
+						// code below.
+						err = c.lowLevelPerform()
+					}
+				}
+			}
+			return err
+		}
+
+		if err != nil && os.IsNotExist(err) {
+			if kind == "symlink" {
+				// For symlinks the path shoud refer to the parent directory.
+				// We set it here so that it gets picked up by the error message
+				// below, in case things fail.
+				path = filepath.Dir(c.Entry.Dir)
+			}
+			// If the element doesn't exist we can attempt to create it.
+			// We will create the parent directory and then the final element
+			// relative to it. The traversed space may be writable so we just
+			// try to create things first.
+			err = tryCreate(createSymlinks)
+
+			if err2, _ := err.(*ReadOnlyFsError); err2 != nil {
+				// If the writing failed because the underlying filesystem
+				// is read-only we can construct a writable mimic to fix that.
+				changes, err = createWritableMimic(err2.Path)
+				if err != nil {
+					return nil, fmt.Errorf("cannot create writable mimic over %q: %s", err2.Path, err)
+				}
+				// We can now try again.
+				err = tryCreate(createSymlinks)
+			}
+			// Check if we eventually succeeded.
+			if err != nil {
+				return changes, err
+			}
+		} else if err != nil {
+			// If we cannot inspect the element let's just bail out.
+			return nil, fmt.Errorf("cannot inspect %q: %v", path, err)
+		} else {
+			// If the element already existed we just need to ensure it is of
+			// the correct type. The desired type depends on the kind of entry
+			// we are working with.
+			switch kind {
+			case "":
+				if !fi.Mode().IsDir() {
+					return nil, fmt.Errorf("cannot use %q for mounting, not a directory", path)
+				}
+			case "file":
+				if !fi.Mode().IsRegular() {
+					return nil, fmt.Errorf("cannot use %q for mounting, not a regular file", path)
+				}
+			case "symlink":
+				// When we want to create a symlink we just need the empty
+				// space so anything that is in the way is a problem.
+				return nil, fmt.Errorf("cannot create symlink in %q, existing file in the way", path)
+			}
 		}
-		// If this is a bind mount then create the source directory as well.
-		// This allows snaps to share a subset of their data easily.
+
+		// At this time we can be sure that the element (for files and
+		// directories) exists and is of the right type or that it (for
+		// symlinks) doesn't exist but the parent directory does.
```

Comment:
Again, only within the context of the snap's processes. We're going to have to think carefully if we want to use sources or destinations that are under the unprivileged, unconfined user's home (or other writable areas) since that could potentially lead to escalation of privileges.